doc/coding_guide/Coding_guidelines.md

```diff
@@ -0,0 +1,650 @@
+
+    Example:
+
+```
+    int array[4] = {1, 2, 3, 4};
+    std::sort(array, array + 4); // correct right bound value is begin + 4 (maybe tricky and lead to errors)
+    std::sort(std::begin(array), std::end(array)); // also works and it is impossible to make mistake in boundaries (it also unified code for any container)
+```
+-   C++17 brings ```std::size()``` and ```std::data()``` functions which are unified interface to get data and size of any container.
+
+-   Use ```std::strings``` for strings.
+
+-   If you need unified callback interface in your class you may use ```std::function``` to represent any callable object (function, lambda). Remember that ```std::function``` brings heap indirection on every call.
+
+-   C++11 provided powerful mechanisms for random numbers generation.
+
+-   C++11 also brought ```std::chrono``` with full functional operations with time and timers.
+
+-   Do not define limits for built in types on your own. You can obtain them from ```std::numeric_limits```.
+
+-   Do not write your implementation if you need to find, sort, move, accumulate (get sum or product) elements in containers. Use ```std::find```/```find_if```/```accumulate```/```count```/```count_if```. Except possible errors in implementation, STL algorithms can use internal threading to get even more speed up (available since C++17).
+
+-   If you need to apply some function to every element of container, use ```std::for_each()```.
+
+-   Be sure to know all algorithms in picture below (they are the most useful, STL has much more of them).
+
+    ![Most suitable STL algorithms](../pic/resized_stl_best.png)
+
+Figure 2. List of most useful STL algorithms (made by Konstantin Vladimirov, see this [link](https://sourceforge.net/projects/cpp-lects-rus/files/slides-cpp/slides-2018/18-algorithms.pdf/download))
+
+**Common mistakes**
+
+19. This is list of some most often mistakes seen in code.
+
+-   Misusage or overusage of references. References are better to use instead of pointers, it allows to skip the pointer check. Some developers consider ```const``` reference as a way to avoid undesired copying of input parameters to functions. The common rule here is that it is very cheap (faster than to dereference a pointer) to copy anything with size less than a cache line (about 64 bytes!).
+
+-   Use prefix ```++i``` operator. Almost all increments which developers use are actually prefix increments. Use postfix ```i++``` only if you need a side effect of postfix ```i++```. This may introduce sufficient difference for some untrivial iterators.
+
+-   Always add ```default:``` in switch operator.
+
+-   If you allocate some buffer during work with C-style strings, do not forget to count terminal ```'\0'``` symbol.
+
+Feel free to improve this document. Bring more examples and rules. But please try to keep size of document small enough. This document is for wide audience of developers, so avoid insufficient details and rare cases.
```

Comment:
Good, but... Sometimes it is too wordy. Hence, too big. I barely made it to line ~100 and at line ~130 I almost stopped looking scrolling down hoping that my eye will catch something. Try to be more precise and better structured... Maybe something like:
1. NO dead code. You don't need it - delete it!
     - Do NOT left commented out code like:
```
//if (x > 5) {
//    printf("DEBUG: x=%d", x);
//}
```

      - Do NOT left code under defines which are never used, typically `#if 0`:

```
#if 0
if (x > 5) {
    printf("DEBUG: x=%d", x);
}
#endif 
```

/sorry for bad formatting of the second block/

Explore whether markdown has some Expand sections. In that way you can hide some explanation details of "why?" you claim this requirement.

Reply:
Yes, I agree that there is still a lot of work to do. Better wording, better structuring, better formatting, more examples... Anyway I think we should proceed with review and merge and apply incremental changes after PR is merged. Because we need to have guidelines document, and I think current state is good enough